File: backend/dispatch/src/main/java/com/commonsengine/dispatch/infrastructure/persistence/DispatchPersistence.java
package com.commonsengine.dispatch.infrastructure.persistence;

import com.commonsengine.dispatch.domain.Model.TaskStatus;
import com.commonsengine.dispatch.domain.Model.TimeSlot;
import com.commonsengine.dispatch.domain.Model.WorkerPreferences;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;

import java.time.Instant;
import java.util.Set;

/**
 * DispatchTaskEntity + WorkerPreferencesEntity + Mappers。
 *
 * 聚合在一个文件以匹配 Kotlin 原结构（DispatchPersistence.kt）。
 * JSON 列（service_types / working_hours）通过 ObjectMapper 序列化。
 */
public final class DispatchPersistence {

    private DispatchPersistence() {
    }

    @Entity
    @Table(name = "dispatch_tasks")
    public static class DispatchTaskEntity {

        @Id
        @GeneratedValue(strategy = GenerationType.UUID)
        @Column(name = "id", columnDefinition = "uuid")
        private String id;

        @Column(name = "service_type", nullable = false)
        private String serviceType;

        @Column(name = "pickup_lat", nullable = false)
        private double pickupLat;

        @Column(name = "pickup_lng", nullable = false)
        private double pickupLng;

        @Column(name = "assigned_worker_id")
        private String assignedWorkerId;

        @Enumerated(EnumType.STRING)
        @Column(name = "status", nullable = false)
        private TaskStatus status;

        @Column(name = "created_at", nullable = false)
        private Instant createdAt;

        public DispatchTaskEntity() {
        }

        public String getId() {
            return id;
        }

        public void setId(String id) {
            this.id = id;
        }

        public String getServiceType() {
            return serviceType;
        }

        public void setServiceType(String serviceType) {
            this.serviceType = serviceType;
        }

        public double getPickupLat() {
            return pickupLat;
        }

        public void setPickupLat(double pickupLat) {
            this.pickupLat = pickupLat;
        }

        public double getPickupLng() {
            return pickupLng;
        }

        public void setPickupLng(double pickupLng) {
            this.pickupLng = pickupLng;
        }

        public String getAssignedWorkerId() {
            return assignedWorkerId;
        }

        public void setAssignedWorkerId(String assignedWorkerId) {
            this.assignedWorkerId = assignedWorkerId;
        }

        public TaskStatus getStatus() {
            return status;
        }

        public void setStatus(TaskStatus status) {
            this.status = status;
        }

        public Instant getCreatedAt() {
            return createdAt;
        }

        public void setCreatedAt(Instant createdAt) {
            this.createdAt = createdAt;
        }
    }

    @Entity
    @Table(name = "worker_preferences")
    public static class WorkerPreferencesEntity {

        @Id
        @Column(name = "worker_id")
        private String workerId;

        @Column(name = "service_types", columnDefinition = "jsonb")
        private String serviceTypesJson;

        @Column(name = "working_hours", columnDefinition = "jsonb")
        private String workingHoursJson;

        @Column(name = "max_service_radius_meters")
        private double maxServiceRadiusMeters;

        public WorkerPreferencesEntity() {
        }

        public String getWorkerId() {
            return workerId;
        }

        public void setWorkerId(String workerId) {
            this.workerId = workerId;
        }

        public String getServiceTypesJson() {
            return serviceTypesJson;
        }

        public void setServiceTypesJson(String serviceTypesJson) {
            this.serviceTypesJson = serviceTypesJson;
        }

        public String getWorkingHoursJson() {
            return workingHoursJson;
        }

        public void setWorkingHoursJson(String workingHoursJson) {
            this.workingHoursJson = workingHoursJson;
        }

        public double getMaxServiceRadiusMeters() {
            return maxServiceRadiusMeters;
        }

        public void setMaxServiceRadiusMeters(double maxServiceRadiusMeters) {
            this.maxServiceRadiusMeters = maxServiceRadiusMeters;
        }
    }

    /** DispatchTask 静态映射器 */
    public static final class DispatchTaskMapper {

        private DispatchTaskMapper() {
        }

        public static com.commonsengine.dispatch.domain.Model.DispatchTask toDomain(DispatchTaskEntity e) {
            if (e == null) {
                return null;
            }
            return new com.commonsengine.dispatch.domain.Model.DispatchTask(
                    e.getId(),
                    e.getServiceType(),
                    e.getPickupLat(),
                    e.getPickupLng(),
                    e.getAssignedWorkerId(),
                    e.getStatus(),
                    e.getCreatedAt()
            );
        }

        public static DispatchTaskEntity toEntity(
                com.commonsengine.dispatch.domain.Model.DispatchTask t) {
            DispatchTaskEntity e = new DispatchTaskEntity();
            e.setId(t.id());
            e.setServiceType(t.serviceType());
            e.setPickupLat(t.pickupLat());
            e.setPickupLng(t.pickupLng());
            e.setAssignedWorkerId(t.assignedWorkerId());
            e.setStatus(t.status());
            e.setCreatedAt(t.createdAt());
            return e;
        }
    }

    /** WorkerPreferences 映射器——依赖 ObjectMapper 处理 JSON 列 */
    public static final class WorkerPreferencesMapper {

        private static final ObjectMapper MAPPER = new ObjectMapper();
        private static final ObjectMapper MAPPER2 = MAPPER; // alias
        private static final TypeReference<Set<String>> STRING_SET = new TypeReference<>() {
        };

        private WorkerPreferencesMapper() {
        }

        public static WorkerPreferences toDomain(WorkerPreferencesEntity e) {
            if (e == null) {
                return null;
            }
            return new WorkerPreferences(
                    e.getWorkerId(),
                    parseStringSet(e.getServiceTypesJson()),
                    parseTimeSlot(e.getWorkingHoursJson()),
                    e.getMaxServiceRadiusMeters()
            );
        }

        public static WorkerPreferencesEntity toEntity(WorkerPreferences p) {
            WorkerPreferencesEntity e = new WorkerPreferencesEntity();
            e.setWorkerId(p.workerId());
            e.setServiceTypesJson(toJson(p.serviceTypes()));
            e.setWorkingHoursJson(toJson(p.workingHours()));
            e.setMaxServiceRadiusMeters(p.maxServiceRadiusMeters());
            return e;
        }

        private static Set<String> parseStringSet(String json) {
            if (json == null || json.isBlank()) {
                return Set.of();
            }
            try {
                return MAPPER.readValue(json, STRING_SET);
            } catch (Exception ex) {
                return Set.of();
            }
        }

        private static TimeSlot parseTimeSlot(String json) {
            if (json == null || json.isBlank()) {
                return null;
            }
            try {
                return MAPPER.readValue(json, TimeSlot.class);
            } catch (Exception ex) {
                return null;
            }
        }

        private static String toJson(Object value) {
            if (value == null) {
                return null;
            }
            try {
                return MAPPER.writeValueAsString(value);
            } catch (Exception ex) {
                return null;
            }
        }
    }
}

File: backend/dispatch/src/main/kotlin/com/commonsengine/dispatch/infrastructure/persistence/DispatchPersistence.kt
package com.commonsengine.dispatch.infrastructure.persistence

import com.commonsengine.dispatch.domain.DispatchTask
import com.commonsengine.dispatch.domain.TimeSlot
import com.commonsengine.dispatch.domain.WorkerPreferences
import com.commonsengine.platform.domain.ServiceType
import com.commonsengine.platform.geo.GeoPoint
import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Table
import java.time.Instant

private val mapper = jacksonObjectMapper()

// ── GeoPoint JSON 序列化辅助 ──────────────────────────

private fun geoPointsToJson(points: List<GeoPoint>): String =
    mapper.writeValueAsString(points.map { mapOf("lat" to it.lat, "lng" to it.lng) })

private fun jsonToGeoPoints(json: String): List<GeoPoint> {
    if (json.isBlank()) return emptyList()
    val type = object : TypeReference<List<Map<String, Double>>>() {}
    return mapper.readValue(json, type).map { GeoPoint(it["lat"]!!, it["lng"]!!) }
}

// ── Enum Set JSON 序列化辅助 ──────────────────────────

private fun enumSetToJson(types: Set<ServiceType>): String =
    if (types.isEmpty()) "" else mapper.writeValueAsString(types.map { it.name })

private fun jsonToServiceTypeSet(json: String?): Set<ServiceType> {
    if (json.isNullOrBlank()) return emptySet()
    val type = object : TypeReference<List<String>>() {}
    return mapper.readValue(json, type).mapNotNull { runCatching { ServiceType.valueOf(it) }.getOrNull() }.toSet()
}

// ── String Set JSON 序列化辅助 ────────────────────────

private fun stringSetToJson(set: Set<String>): String =
    if (set.isEmpty()) "" else mapper.writeValueAsString(set)

private fun jsonToStringSet(json: String?): Set<String> {
    if (json.isNullOrBlank()) return emptySet()
    val type = object : TypeReference<List<String>>() {}
    return mapper.readValue(json, type).toSet()
}

// ── TimeSlot Set JSON 序列化辅助 ──────────────────────

private fun timeSlotsToJson(slots: Set<TimeSlot>): String =
    if (slots.isEmpty()) "" else mapper.writeValueAsString(slots.map {
        mapOf("dayOfWeek" to it.dayOfWeek, "startHour" to it.startHour, "endHour" to it.endHour)
    })

private fun jsonToTimeSlots(json: String?): Set<TimeSlot> {
    if (json.isNullOrBlank()) return emptySet()
    val type = object : TypeReference<List<Map<String, Int>>>() {}
    return mapper.readValue(json, type).map {
        TimeSlot(it["dayOfWeek"]!!, it["startHour"]!!, it["endHour"]!!)
    }.toSet()
}

// ══════════════════════════════════════════════════════
// DispatchTaskEntity
// ══════════════════════════════════════════════════════

/**
 * 调度任务 JPA 实体
 *
 * pickups/dropoffs: List<GeoPoint> 以 JSON 字符串存储。
 */
@Entity
@Table(name = "dispatch_tasks")
class DispatchTaskEntity(

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long = 0,

    @Column(name = "task_id", nullable = false, length = 36)
    val taskId: String,

    @Column(name = "worker_id", nullable = false, length = 36)
    val workerId: String,

    @Enumerated(EnumType.STRING)
    @Column(name = "service_type", nullable = false, length = 30)
    val serviceType: ServiceType,

    /** JSON: [{"lat":..,"lng":..}] */
    @Column(name = "pickups", nullable = false, columnDefinition = "TEXT")
    val pickups: String,

    /** JSON: [{"lat":..,"lng":..}] */
    @Column(name = "dropoffs", nullable = false, columnDefinition = "TEXT")
    val dropoffs: String,

    @Column(name = "estimated_distance_meters", nullable = false)
    val estimatedDistanceMeters: Double = 0.0,

    @Column(name = "estimated_duration_minutes", nullable = false)
    val estimatedDurationMinutes: Int = 0,

    @Column(name = "assigned_at", nullable = false)
    val assignedAt: Instant,

    @Column(name = "deadline")
    val deadline: Instant? = null,
)

// ── DispatchTask Entity ↔ Domain ──────────────────────

fun DispatchTaskEntity.toDomain(): DispatchTask = DispatchTask(
    id = taskId,
    workerId = workerId,
    serviceType = serviceType,
    pickups = jsonToGeoPoints(pickups),
    dropoffs = jsonToGeoPoints(dropoffs),
    estimatedDistanceMeters = estimatedDistanceMeters,
    estimatedDurationMinutes = estimatedDurationMinutes,
    assignedAt = assignedAt,
    deadline = deadline,
)

fun DispatchTask.toEntity(): DispatchTaskEntity = DispatchTaskEntity(
    taskId = id,
    workerId = workerId,
    serviceType = serviceType,
    pickups = geoPointsToJson(pickups),
    dropoffs = geoPointsToJson(dropoffs),
    estimatedDistanceMeters = estimatedDistanceMeters,
    estimatedDurationMinutes = estimatedDurationMinutes,
    assignedAt = assignedAt,
    deadline = deadline,
)

// ══════════════════════════════════════════════════════
// WorkerPreferencesEntity
// ══════════════════════════════════════════════════════

/**
 * 劳动者工作偏好 JPA 实体
 *
 * 反榨取设计核心数据：劳动者自己设定的工作边界，引擎必须尊重。
 * 所有 Set 类型以 JSON 字符串存储。
 */
@Entity
@Table(name = "worker_preferences")
class WorkerPreferencesEntity(

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long = 0,

    @Column(name = "worker_id", nullable = false, length = 36)
    val workerId: String,

    /** JSON: ["RIDE_HAILING","FOOD_DELIVERY"] */
    @Column(name = "preferred_service_types", columnDefinition = "TEXT")
    val preferredServiceTypes: String? = null,

    /** JSON: ["chengdu_wuhou", ...] */
    @Column(name = "preferred_regions", columnDefinition = "TEXT")
    val preferredRegions: String? = null,

    /** JSON: ["chengdu_jinjiang", ...] */
    @Column(name = "excluded_regions", columnDefinition = "TEXT")
    val excludedRegions: String? = null,

    /** JSON: [{"dayOfWeek":1,"startHour":8,"endHour":12}, ...] */
    @Column(name = "preferred_time_slots", columnDefinition = "TEXT")
    val preferredTimeSlots: String? = null,

    /** JSON: [{"dayOfWeek":7,"startHour":0,"endHour":6}, ...] */
    @Column(name = "excluded_time_slots", columnDefinition = "TEXT")
    val excludedTimeSlots: String? = null,

    @Column(name = "max_concurrent_orders", nullable = false)
    val maxConcurrentOrders: Int = 3,

    @Column(name = "max_daily_hours", nullable = false)
    val maxDailyHours: Double = 12.0,
)

// ── WorkerPreferences Entity ↔ Domain ─────────────────

fun WorkerPreferencesEntity.toDomain(): WorkerPreferences = WorkerPreferences(
    workerId = workerId,
    preferredServiceTypes = jsonToServiceTypeSet(preferredServiceTypes),
    preferredRegions = jsonToStringSet(preferredRegions),
    excludedRegions = jsonToStringSet(excludedRegions),
    preferredTimeSlots = jsonToTimeSlots(preferredTimeSlots),
    excludedTimeSlots = jsonToTimeSlots(excludedTimeSlots),
    maxConcurrentOrders = maxConcurrentOrders,
    maxDailyHours = maxDailyHours,
)

fun WorkerPreferences.toEntity(): WorkerPreferencesEntity = WorkerPreferencesEntity(
    workerId = workerId,
    preferredServiceTypes = enumSetToJson(preferredServiceTypes),
    preferredRegions = stringSetToJson(preferredRegions),
    excludedRegions = stringSetToJson(excludedRegions),
    preferredTimeSlots = timeSlotsToJson(preferredTimeSlots),
    excludedTimeSlots = timeSlotsToJson(excludedTimeSlots),
    maxConcurrentOrders = maxConcurrentOrders,
    maxDailyHours = maxDailyHours,
)
